### ISSUE TITLE
"Field" syntax is consumed by OTel collector config loading process

### ISSUE THREAD

**Issue by djaglowski** [open-telemetry/opentelemetry-log-collection#164]
The OTel collector [automatically expands](https://github.com/open-telemetry/opentelemetry-collector/blob/df76aa36cd124ba4ae1b3d1ac1941155f86ea568/config/configloader/config.go#L515) tokens that appear to be environment variables. (i.e. `$hostname`)

This library's [field](https://github.com/open-telemetry/opentelemetry-log-collection/blob/main/docs/types/field.md#fields) syntax relies heavily on usage of `$resources`, `$attributes`, and `$body`. 

The documented approach to specifying fields does not work. Instead, the `$resources`, `$attributes`, and `$body` tokens are expanded (typically replaced with an empty string).

It is possible to escape the expansion (`$$attributes`), but as a component of the open telemetry project, we would ideally not require users to worry about this.

Fortunately, this usage of `$` in this library was a largely arbitrary decision. I suggest that we should choose an alternate character or mechanism for field syntax.

**Comment by pmalek-sumo** [comment]
I wanted to open an new issue but I thought this is connected somehow:

So my config:

```yaml
  filelog:
    include: [ "**.log" ]
    operators:
      - type: metadata
        id: metadata1
        resource:
          ttt: 'EXPR( $ == "x" ? "asd" : "qwe" )'
      - type: metadata
        id: metadata2
        resource:
          resource2: $resource
      - type: metadata
        id: metadata3
        attributes:
          attribute2: $attributes
```

producing (empty resource and attributes) entries:

```
2021-07-13T18:50:21.161+0200    INFO    loggingexporter/logging_exporter.go:135 LogsExporter    {"#logs": 1}
2021-07-13T18:50:21.161+0200    DEBUG   loggingexporter/logging_exporter.go:141 ResourceLog #0
Resource labels:
     -> ttt: STRING(qwe)
     -> resource2: STRING()
InstrumentationLibraryLogs #0
InstrumentationLibrary
LogRecord #0
Timestamp: 2021-07-13 16:50:21.062134 +0000 UTC
Severity: Undefined
ShortName:
Body: 2021-07-13T18:50:20.995+0200      debug   sumologicextension@v0.27.0/extension.go:281     Heartbeat sent  {"kind": "extension", "name": "sumologic/1"}
Attributes:
     -> file.name: STRING(reg.log)
     -> attribute2: STRING()
```

or verbatim text with escaped `$` when replacing `$resource` -> `$$resource` and `$attributes` -> `$$attributes` in the config

```
     -> resource2: STRING($resource)
...
     -> attribute2: STRING($attributes)
...
```

is the consequence of this issue or am I not understanding how the expression engine works here?

IMO this should be better explained in here https://github.com/open-telemetry/opentelemetry-log-collection/blob/main/docs/types/expression.md

**Comment by djaglowski** [comment]
Yes, unfortunately any `$` in the collector config will be replaced, so users currently need to use `$$` in place of `$`, regardless of the context in which it is used. This will also affect usage of `$` in regexes, for example.

Given that this can occur in a variety of situations, I think we should have a standalone document that is linked to wherever `$` is used in documentation, examples, etc.

What I'm not sure about, is whether we should update all examples to actually show `$$` instead of `$`. It's not ideal that the user would have to determine that escaping is necessary, but it's also possible that examples with escaping built in may cause confusion about whether or not escaping is still necessary. In theory, there are some contexts in which showing `$$` would be inaccurate, such as within plugins.

As a side note, in this expression `'EXPR( $ == "x" ? "asd" : "qwe" )'`, the usage of `$` is intended as shorthand for `$body`. This is technically supported, but I've stripped all references to this shorthand notation from documentation, in the hopes of encouraging users to use the more verbose `$body`, which as you've noted should actually be `$$body` when used in the collector config.

**Comment by djaglowski** [comment]
Copying [@tigrannajaryan's comment from #410](https://github.com/open-telemetry/opentelemetry-log-collection/issues/410#issuecomment-1061185076), as well as my initial response:

> Somewhat related question: should we use CamelCase for referencing built-in top-level fields, e.g. Attributes, Resource, SeverityNumber, etc?
>
> I was envisioning a future language where a short-hand syntax for referencing LogRecord attributes is simply their names provided that they start with lower case character and it will be easy to distinguish them from top-level fields because of capitalization rules (Otel conventions all use lower case for attribute names, although not mandated). We don't need to do this now, but going CamelCase for builtin fields could keep this possibility open.
>
> CamelCase also coincides with how our data model is specified.

I like that it would be very clear that you are referring to a top-level field.

On the other hand, I don't necessarily think it's a great idea to have an implicit/default top-level field. Maybe what you're thinking about is not necessarily for this codebase specifically, but just pointing out that in retrospect it didn't really save much trouble but definitely caused some later.

**Comment by djaglowski** [comment]
Having thought about this some more, I propose that we stick with the syntax proposed in #364, which is simple and clear. 

I think switching to CamelCase w/ or w/o a default top-level field unnecessarily complicates the syntax.

**Comment by djaglowski** [comment]
Resolved by #364
